AFQ/models/msmt.py

```diff
@@ -0,0 +1,515 @@
+import multiprocessing
+import warnings
+import numpy as np
+from tqdm import tqdm
+import ray
+
+from scipy.optimize import minimize
+from scipy.sparse import csr_matrix
+
+from numba import njit, prange, set_num_threads, config
+from numba.core.errors import NumbaPerformanceWarning
+
+import osqp
+
+from dipy.reconst.mcsd import MSDeconvFit
+from dipy.reconst.mcsd import MultiShellDeconvModel
+
+from AFQ.utils.stats import chunk_indices
+
+
+__all__ = ["fit"]
+
+
+# CPU implementation of primal-dual Interior Point method
+# for convex quadratic constrained optimization (QP)
+# Specifically, ||Rx-d||_2 where Ax>=b
+# parallelized to solve 10s-100s of thousands of QPs
+# ultimately used to fit MSMT CSD
+@njit(fastmath=True)
+def solve_qp(Rt, R_pinv, G, A, At, A_outer, b, x0,
+             d, max_iter, tol):
+    '''
+    Solves 1/2*x^t*G*x+(Rt*d)^t*x given Ax>=b
+    In MSMT, G, R, A, b are the same across voxels,
+    but there are different d for different voxels.
+    This fact is not used currently. So this is a more general
+    batched CLS QP solver.
+
+    Let:
+    c=(Rt*d)^t
+    L = diag(l)
+    Y = diag(y)
+    mu as centering parameter that tends to 0 with iteration number.
+
+    Set up with interior points, this is reformulated to:
+    | G  0 -A.T | |dx|   |0 |   | G*x-A.T*l+c |
+    | A -I  0   | |dy| = |0 | - | A*x-y-b     |
+    | 0  L  Y   | |dl|   |mu|   | YL          |
+
+    This reduces to:
+    |G -A.T| |dx| = | -G*x-A.T*l+c          |
+    |A Y\L | |dl|   | -(A*x-y-b) + (-y+mu/l)|
```

Comment:
The `\L` on this line is raising an error in the documentation build, because Sphinx uses the `\` character as an escape character. Could we change this notation? Is this equivalent to something like `pinv(Y) @ L`? I see this a couple of times below in the docstring as well.